File: tools/zetl-ast-py/src/zetl_ast/walk.py
"""
Low-level traversal helpers.

``walk()`` is a generator; ``map_nodes()`` does one-pass in-place
replacement. Prefer :func:`zetl_ast.dispatch` for 80% of hook work —
``walk`` is the building-block when you need to collect or filter.
"""

from __future__ import annotations

from typing import Any, Callable, Iterable, Iterator, List, Optional, Union

from .ast import Node

_CHILD_BEARING = frozenset(
    {
        "Document",
        "Paragraph",
        "Heading",
        "Emphasis",
        "Strong",
        "Link",
        "Image",
        "BlockQuote",
        "ListItem",
        "List",
    }
)


def children_of(node: Node) -> Iterable[Node]:
    """Child list for ``node`` in document order. ``()`` for leaves."""
    if node.get("type") in _CHILD_BEARING:
        return node.get("children") or ()
    return ()


def _matcher(
    filter_: Union[str, Iterable[str], None],
) -> Callable[[str], bool]:
    if filter_ is None:
        return lambda _t: True
    if isinstance(filter_, str):
        target = filter_
        return lambda t: t == target
    allowed = frozenset(filter_)
    return lambda t: t in allowed


def walk(
    root: Node,
    *,
    type: Union[str, Iterable[str], None] = None,
) -> Iterator[Node]:
    """
    Depth-first pre-order traversal.

    Yields ``root`` itself, then each child in order, recursing into
    child-bearing nodes. Skips leaves (``Text``, ``Code``, ``Wikilink``,
    etc.).

    If ``type`` is given, yields only matching nodes — either a single
    node-name constant or any iterable of them.
    """
    match = _matcher(type)
    yield from _walk_impl(root, match)


def _walk_impl(node: Node, match: Callable[[str], bool]) -> Iterator[Node]:
    if match(node.get("type", "")):
        yield node
    for child in children_of(node):
        yield from _walk_impl(child, match)


def map_nodes(
    root: Node,
    replacer: Callable[[Node], Optional[Node]],
) -> Node:
    """
    One-pass in-place replacement.

    Calls ``replacer(node)`` on every child. A returned node replaces
    the original; ``None`` (or the same object) leaves it unchanged.
    Descends into the replacement, not the original.
    """
    children = _mutable_children(root)
    if children is not None:
        for i, curr in enumerate(children):
            nxt = replacer(curr)
            if nxt is not None and nxt is not curr:
                children[i] = nxt
                map_nodes(nxt, replacer)
            else:
                map_nodes(curr, replacer)
    return root


def _mutable_children(node: Node) -> Optional[List[Any]]:
    if node.get("type") in _CHILD_BEARING:
        kids = node.get("children")
        if isinstance(kids, list):
            return kids
    return None

File: tools/ztl-ast-js/src/walk.ts
// walk() — generator-based traversal.
//
// CON-3210 sample:
//   for (const node of walk(ast, { type: BlockQuote })) { ... }
//
// The low-level API; `dispatch()` is the recommended 80% path (REQ-3218).

import type { AnyNode, BlockNode, DocumentNode, InlineNode, ListItemNode, NodeOf, NodeType } from "./ast.ts";

export interface WalkOptions<T extends NodeType = NodeType> {
    /** Restrict the generator to nodes whose `type` matches. */
    type?: T | readonly T[];
}

/**
 * Depth-first pre-order traversal. Visits the root itself, then each
 * child in order, recursing into children-bearing nodes. Skips inside
 * `CodeBlock`, `HtmlBlock`, `HtmlInline`, `SplBlock`, etc. since those
 * have no child nodes.
 */
export function walk<T extends NodeType = NodeType>(
    root: AnyNode,
    opts?: WalkOptions<T>,
): Generator<NodeOf<T>> {
    const filter = opts?.type;
    const matcher = buildMatcher(filter);
    return walkImpl(root, matcher) as Generator<NodeOf<T>>;
}

function* walkImpl(node: AnyNode, matcher: (t: string) => boolean): Generator<AnyNode> {
    if (matcher(node.type)) yield node;
    const children = childrenOf(node);
    for (const child of children) {
        yield* walkImpl(child, matcher);
    }
}

function buildMatcher(filter: NodeType | readonly NodeType[] | undefined): (t: string) => boolean {
    if (filter === undefined) return () => true;
    if (Array.isArray(filter)) {
        const set = new Set<string>(filter as readonly string[]);
        return (t) => set.has(t);
    }
    return (t) => t === (filter as string);
}

/**
 * Child list for a node in document order. Returns `[]` for leaf nodes
 * (Text, Code, ThematicBreak, LineBreak, Embed, Wikilink, ...).
 */
export function childrenOf(node: AnyNode): ReadonlyArray<AnyNode> {
    switch (node.type) {
        case "Document":
        case "Paragraph":
        case "Heading":
        case "Emphasis":
        case "Strong":
        case "Link":
        case "Image":
            return (node as DocumentNode | ParagraphLike).children;
        case "BlockQuote":
        case "ListItem":
            return (node as BlockQuoteLike).children;
        case "List":
            return (node as ListLike).children;
        default:
            return [];
    }
}

// Local structural aliases keep the switch readable without depending
// on the full discriminated union.
type ParagraphLike = { children: InlineNode[] };
type BlockQuoteLike = { children: BlockNode[] };
type ListLike = { children: ListItemNode[] };

/**
 * In-place replacement of every node for which `predicate` returns a
 * new node. The input tree is mutated; `null`/`undefined` returns leave
 * the node unchanged. Runs one pass (does not re-visit substitutes).
 *
 * Exposed as a building-block for dispatch handlers that want to swap
 * a subtree without walking manually.
 */
export function mapNodes(
    root: AnyNode,
    replacer: (node: AnyNode) => AnyNode | null | undefined,
): AnyNode {
    const children = mutableChildren(root);
    if (children) {
        for (let i = 0; i < children.length; i++) {
            const curr = children[i]!;
            const next = replacer(curr);
            if (next != null && next !== curr) {
                children[i] = next as never;
                // Descend into the replacement, not the original.
                mapNodes(next, replacer);
            } else {
                mapNodes(curr, replacer);
            }
        }
    }
    return root;
}

function mutableChildren(node: AnyNode): AnyNode[] | null {
    switch (node.type) {
        case "Document":
        case "Paragraph":
        case "Heading":
        case "Emphasis":
        case "Strong":
        case "Link":
        case "Image":
        case "BlockQuote":
        case "ListItem":
        case "List":
            // Narrowed at callsite — the schema guarantees `children` exists.
            return (node as { children: AnyNode[] }).children;
        default:
            return null;
    }
}
